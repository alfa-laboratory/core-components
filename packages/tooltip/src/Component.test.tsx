import React from 'react';
import {
    render,
    fireEvent,
    act,
    RenderResult,
    waitForElement,
    waitForElementToBeRemoved,
} from '@testing-library/react';

import { Tooltip, TooltipProps } from './index';

jest.mock('react-transition-group', () => {
    return { Transition: jest.fn(props => (props.in ? props.children(props) : null)) };
});

const renderTooltip = async (props: TooltipProps): Promise<RenderResult> => {
    let result;

    await act(async () => {
        result = await render(<Tooltip {...props}>{props.children}</Tooltip>);
    });

    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore
    return result;
};

describe('Display tests', () => {
    it('should display correctly', async () => {
        const component = await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
        });

        expect(component).toMatchSnapshot();
    });

    it('should display correctly if tooltip open', async () => {
        const component = await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
            open: true,
        });

        expect(component).toMatchSnapshot();
    });
});

describe('Attributes tests', () => {
    it('should set data-test-id attribute', async () => {
        const testId = 'tooltip';
        const className = 'custom';

        await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
            open: true,
            contentClassName: className,
            dataTestId: testId,
        });

        const tooltipContentWrap = document.querySelector(`.${className}`);

        const testIdAttr = tooltipContentWrap?.getAttribute('data-test-id');

        expect(testIdAttr).toBe(testId);
    });
});

describe('Styles tests', () => {
    it('should set custom css class to content component', async () => {
        const className = 'custom';
        const testId = 'tooltip';

        const { getByTestId } = await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
            open: true,
            contentClassName: className,
            dataTestId: testId,
        });

        expect(getByTestId(testId)).toHaveClass(className);
    });

    it('should set custom css class to arrow component', async () => {
        const arrowClassName = 'custom-arrow-class';

        await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
            open: true,
            arrowClassName,
        });

        expect(document.body.querySelector(`.${arrowClassName}`)).toHaveClass('arrow');
    });
});

describe('Render tests', () => {
    it('should unmount without errors', async () => {
        const { unmount } = await renderTooltip({
            children: <div>Hover me</div>,
            content: <div>I am tooltip</div>,
            open: true,
        });

        expect(unmount).not.toThrowError();
    });
});

describe('Click event tests', () => {
    it('should open/close if click on children', async () => {
        const childrenText = 'Click me';
        const contentText = 'I am tooltip';
        const onClose = jest.fn();
        const onOpen = jest.fn();

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div>{contentText}</div>,
            trigger: 'click',
            onClose,
            onOpen,
        });

        const children = getByText(childrenText);

        fireEvent.click(children);

        const content = await waitForElement(() => getByText(contentText));

        expect(content).toBeInTheDocument();

        fireEvent.click(children);

        expect(content).not.toBeInTheDocument();

        expect(onOpen).toBeCalledTimes(1);
        expect(onClose).toBeCalledTimes(1);
    });

    it('should close if click outside', async () => {
        const childrenText = 'Click me';
        const contentText = 'I am tooltip';

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div>{contentText}</div>,
            trigger: 'click',
        });

        const children = getByText(childrenText);

        fireEvent.click(children);

        const content = await waitForElement(() => getByText(contentText));

        expect(content).toBeInTheDocument();

        fireEvent.click(document.body);

        expect(content).not.toBeInTheDocument();
    });

    it('should stay open if prop `open` is `true`', async () => {
        const childrenText = 'Click me';
        const contentText = 'I am tooltip';

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div>{contentText}</div>,
            trigger: 'click',
            open: true,
        });

        const content = getByText(contentText);
        const children = getByText(childrenText);

        expect(content).toBeInTheDocument();

        fireEvent.click(document.body);

        expect(content).toBeInTheDocument();

        fireEvent.click(children);

        expect(content).toBeInTheDocument();
    });

    it('should stay close if prop `open` is `false`', async () => {
        const childrenText = 'Click me';
        const contentText = 'I am tooltip';

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div id='content'>{contentText}</div>,
            trigger: 'click',
            open: false,
        });

        const content = document.querySelector('#content');
        const children = getByText(childrenText);

        expect(content).toBeNull();

        fireEvent.click(children);

        expect(content).toBeNull();
    });
});

describe('Hover event tests', () => {
    it('should open/close if hover on children', async () => {
        const childrenText = 'Click me';
        const contentText = 'I am tooltip';
        const onClose = jest.fn();
        const onOpen = jest.fn();

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div id='content'>{contentText}</div>,
            onClose,
            onOpen,
        });

        const children = getByText(childrenText);

        fireEvent.mouseOver(children);

        const content = await waitForElement(() => getByText(contentText));

        expect(content).toBeInTheDocument();

        fireEvent.mouseOut(children);

        await waitForElementToBeRemoved(() => getByText(contentText));

        expect(content).not.toBeInTheDocument();

        expect(onOpen).toBeCalledTimes(1);
        expect(onClose).toBeCalledTimes(1);
    });

    it('should stay open if prop `open` is `true`', async () => {
        jest.useFakeTimers();

        const childrenText = 'Click me';
        const contentText = 'I am tooltip';
        const onClose = jest.fn();
        const onOpen = jest.fn();

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div>{contentText}</div>,
            onClose,
            onOpen,
            open: true,
        });

        const children = getByText(childrenText);

        fireEvent.mouseOver(children);
        fireEvent.mouseOut(children);

        const content = getByText(contentText);

        expect(content).toBeInTheDocument();
    });

    it('should stay close if prop `open` is `false`', async () => {
        jest.useFakeTimers();

        const childrenText = 'Click me';
        const contentText = 'I am tooltip';
        const onClose = jest.fn();
        const onOpen = jest.fn();

        const { getByText } = await renderTooltip({
            children: <div>{childrenText}</div>,
            content: <div id='content'>{contentText}</div>,
            onClose,
            onOpen,
            open: false,
        });

        const children = getByText(childrenText);

        fireEvent.mouseOver(children);
        fireEvent.mouseOut(children);

        const content = document.querySelector('#content');

        expect(content).toBeNull();
    });
});

describe('Child render tests', () => {
    it('should pass child classname', async () => {
        const childClassName = 'child-classname';
        const testId = 'test-id';

        const { getByTestId } = await renderTooltip({
            children: (
                <div className={childClassName} data-test-id={testId}>
                    Hover me
                </div>
            ),
            content: <div>I am tooltip</div>,
            open: true,
        });

        expect(getByTestId(testId)).toHaveClass(childClassName);
    });
});
