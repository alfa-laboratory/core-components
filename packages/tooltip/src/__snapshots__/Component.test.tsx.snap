// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display tests should display correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="target"
      >
        Hover me
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="target"
    >
      Hover me
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Display tests should display correctly if tooltip open 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="target"
      >
        Hover me
      </div>
    </div>
    <div
      alfa-portal-container=""
    >
      <div
        class="component [object Object] mountOnEnter unmountOnExit in timeout children popper"
        data-popper-escaped="true"
        data-popper-placement="left"
        data-popper-reference-hidden="true"
        style="position: absolute; left: 0px; top: 0px; transition-duration: 300ms; right: 0px; transform: translate(-10px, 0px);"
      >
        <div
          class="component"
        >
          <div>
            I am tooltip
          </div>
        </div>
        <div
          class="arrow"
          style="position: absolute; top: 0px; transform: translate(0px, 0px);"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="target"
    >
      Hover me
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
